### PR TITLE
- Use a unitialized buffer to store hash keys and values.

### PR DESCRIPTION
**Pull request by baldersheim**
- Use placement new and explicit destructor calls to not take the cost before necessary.
- This removes the need for a default constructor and unnecessary initialization.

@havardpe PR

### CODE REVIEW

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 1]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
-    hash_node &operator=(hash_node &&) noexcept = default;
-    hash_node(const hash_node &) = default;             // These will not be created
-    hash_node &operator=(const hash_node &) = default;  // if V is non-copyable.
+        : _node(),
+          _next(next)
+    {
+        new (_node) V(std::move(node));
+    }
+    hash_node(hash_node && rhs) noexcept
+        : hash_node(std::move(rhs.getValue()), rhs._next)
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:118]: need to check 'valid' here as well

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 2]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
+    hash_node &operator=(const hash_node & rhs) {
+        destruct();
+        if (rhs.valid()) {
+            new (_node) V(rhs.getValue());
+            _next = rhs._next;
+        }
+        return *this;
+    }
+    ~hash_node() {
+        if (valid()) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:145]: consider adding outer (static) check `if (!can_skip_destruction<T>::value) {`

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 3]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
     void setNext(next_t next)  { _next = next; }
-    void invalidate()          { _next = invalid; _node = V(); }
+    void invalidate()          { destruct(); }
     void terminate()           { _next = npos; }
     bool valid()         const { return _next != invalid; }
     bool hasNext()       const { return valid() && (_next != npos); }
 private:
-    V       _node;
+    void destruct() {
+        if (valid()) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:162]: consider adding outer (static) check `if (!can_skip_destruction<T>::value) {`

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 4]
```diff
@@ -103,26 +103,68 @@ class hash_node {
     enum {npos=-1u, invalid=-2u};
     hash_node() : _node(), _next(invalid) {}
     hash_node(const V & node, next_t next=npos)
-        : _node(node), _next(next) {}
+        : _node(),
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:106]: I think we can skip 'node' double-initialization by not initializing 'node' here, not having to rely too heavily on compiler optimizations.

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 5]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
+    {
+        new (_node) V(node);
+    }
     hash_node(V &&node, next_t next=npos)
-        : _node(std::move(node)), _next(next) {}
-    hash_node(hash_node &&) noexcept = default;
-    hash_node &operator=(hash_node &&) noexcept = default;
-    hash_node(const hash_node &) = default;             // These will not be created
-    hash_node &operator=(const hash_node &) = default;  // if V is non-copyable.
+        : _node(),
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:112]: I think we can skip 'node' double-initialization by not initializing 'node' here, not having to rely too heavily on compiler optimizations.

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 6]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
+    hash_node &operator=(hash_node && rhs) noexcept {
+        destruct();
+        if (rhs.valid()) {
+            new (_node) V(std::move(rhs.getValue()));
+            _next = rhs._next;
+        }
+        return *this;
+    }
+    hash_node(const hash_node & rhs)
+        : _node(),
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:129]: I think we can skip 'node' double-initialization by not initializing 'node' here, not having to rely too heavily on compiler optimizations.

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 7]
```diff
@@ -103,26 +103,68 @@ class hash_node {
...
+        return *this;
+    }
+    ~hash_node() {
+        if (valid()) {
+            getValue().~V();
+        }
+    }
     bool operator == (const hash_node & rhs) const {
-        return (_next == rhs._next) && (_node == rhs._node);
+        return (_next == rhs._next) && (!valid() || (getValue() == rhs.getValue()));
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:150]: do we really need the == operator?

**Review by havardpe — COMMENTED** [review]
could probably also make faster move/copy by checking for trivial copyability

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 1 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:118]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 6 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:129]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 5 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:112]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 4 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:106]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 2 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:145]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 3 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:162]: Here it must be an inner check as _next must be invalid when leaving this method.

Comment (havardpe) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:162]: or maybe switch things around a bit. The destruct function is not used by the destructor and it does invalidation, which the invalidate could also be responsible for...

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 7 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h:154]: Yes, unfortunately it seems like a requirement for Array class for now as it uses explicit instantiatio.

**Comment by baldersheim** [comment]
@havardpe PTAL

**Comment by havardpe** [comment]
13:04:57 The following tests FAILED:
13:04:57 	369 - searchlib_attributemanager_test_app (Failed)
13:04:57 	390 - searchlib_attribute_searchable_adapter_test_app (Child aborted)
13:04:57 	391 - searchlib_attribute_weighted_set_blueprint_test_app (Child aborted)
13:04:57 	392 - searchlib_attribute_blueprint_test_app (Child aborted)
13:04:57 	590 - staging_vespalib_lrucache_test_app (Failed)
13:04:57 	736 - vespalib_hash_test_app (Failed)

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 8]
```diff
@@ -103,26 +104,73 @@ class hash_node {
...
+    hash_node &operator=(const hash_node & rhs) {
+        destruct();
+        if (rhs.valid()) {
+            new (_node) V(rhs.getValue());
+            _next = rhs._next;
+        }
+        return *this;
+    }
+    ~hash_node() {
+        if (can_skip_destruction<V>::value) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:147]: this needs to be inverted

File: vespalib/src/vespa/vespalib/stllike/hashtable.h [hunk 9]
```diff
@@ -103,26 +104,73 @@ class hash_node {
...
-    void invalidate()          { _next = invalid; _node = V(); }
+    void invalidate()          { destruct(); }
     void terminate()           { _next = npos; }
     bool valid()         const { return _next != invalid; }
     bool hasNext()       const { return valid() && (_next != npos); }
 private:
-    V       _node;
+    void destruct() {
+        if (valid()) {
+            if (can_skip_destruction<V>::value) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/stllike/hashtable.h:167]: this needs to be inverted

**Review by havardpe — APPROVED** [review]

**Comment by vekterli** [comment]
Should `if (!can_skip_destruction<V>::value)` be `constexpr` to not depend on dead code elimination to remove any fluff?

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 8 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h]: Fixed

File: vespalib/src/vespa/vespalib/stllike/hashtable.h (hunk 9 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/stllike/hashtable.h]: Fixed

**Comment by baldersheim** [comment]
@vekterli As far as I can tell this resolves to
  template<typename _Tp, _Tp __v>
    constexpr _Tp integral_constant<_Tp, __v>::value;
in type_traits in stl which I think resolves your concern.

**Comment by vekterli** [comment]
I was thinking in terms of replacing it with `if constexpr (!can_skip_destruction<V>::value)` to avoid even instantiating any destructors in the cases where they're not needed

**Comment by baldersheim** [comment]
Point taken, done

**Pull request merged**